### PR TITLE
Issue 549 - format!-like macros formating

### PR DESCRIPTION
**Pull request by bekh6ex**
Format `format!`,`format_args!`, `panic!`, `print!`, `println!` calls as described in https://github.com/rust-lang-nursery/rustfmt/issues/549

### CODE REVIEW

**Comment by bekh6ex** [comment]
Self tests still are failing. I thin it's better to reformat whole repository after someone reviews the code.

**Comment by nrc** [comment]
Rather than white-listing macros, I think it would be better to try and identify any macro that looks function-like and starts with a string literal as a candidate for this formatting. Does that seem reasonable?

File: src/expr.rs [hunk 1]
```diff
@@ -816,10 +814,8 @@ fn rewrite_if_else(context: &RewriteContext,
         _ => " ",
     };
     let mut result = format!("if{}{}{}{}",
-                             between_if_cond_comment.as_ref().map_or(" ", |str| &**str),
-                             pat_expr_string,
-                             after_cond_comment.as_ref().map_or(after_sep, |str| &**str),
-                             if_block_string);
+                             between_if_cond_comment.as_ref().map_or(" ", |str| &**str), pat_expr_string,
+                             after_cond_comment.as_ref().map_or(after_sep, |str| &**str), if_block_string);
```
Comment (nrc) [inline src/expr.rs]: if you do need to break lines in the arguments, they should then be one arg per line, rather than many args per line

File: src/expr.rs [hunk 2]
```diff
@@ -719,11 +722,7 @@ impl<'a> Rewrite for Loop<'a> {
             .rewrite(context, block_width, offset)
             .map(|result| {
                 format!("{}{}{}{}{}",
-                        label_string,
-                        self.keyword,
-                        pat_expr_string,
-                        block_sep,
-                        result)
+                        label_string, self.keyword, pat_expr_string, block_sep, result)
```
Comment (nrc) [inline src/expr.rs]: Request for future PRs - please could you keep reformatting and new code in different commits please? It makes the new code easier to review.

File: src/lists.rs [hunk 3]
```diff
@@ -28,6 +28,10 @@ pub enum ListTactic {
     Horizontal,
     // Try Horizontal layout, if that fails then vertical.
     HorizontalVertical,
+    // Try Horizontal layout, if that fails then put first element on the first line,
+    // and others as many items as possible per row over (possibly) many rows.
+    // Used for macro calls like `format!`, `println!`, etc
+    FormatLikeMacroCall(usize),
```
Comment (nrc) [inline src/lists.rs]: I don't think it is quite right to add this variant to this enum - one could imagine combining any of these formatting styles with the format-macro style of putting the string lit on its own line. I.e., this should probably be a separate flag into the list formatting functions.

Comment (nrc) [inline src/lists.rs]: What is the `usize` arg for?

File: src/expr.rs (hunk 1 above)
Comment (bekh6ex) [inline src/expr.rs]: Then I didn't understand how should we format `format!` calls.

We can do like this:
  * If fits single line - do single line
  * If doesn't fit single line - put first argument on first line, and all others on the second
  * If still doesn't fit - do vertical layout

Does it seem correct?

**Comment by bekh6ex** [comment]
About white listing.
If we try to guess, then we can end up with randomish formatting.
If we will try to determine macro based on the fact that first argument is static string, then this
```
some_macro!("arg1", "arg2", "arg3");
```
will result in
```
some_macro!("arg1", 
            "arg2", "arg3");
```
which, in my opinion, is pretty strange.
And also if author refactors it a bit, we will have
```
let a1 = "arg1";
some_macro!(a1, 
            "arg2", 
            "arg3");
```
In my opinion, this kind of behaviour is not very obvious to user, and also it would be hard to describe if someone wanted to document it (as a project's code style, for example).

If we try to detect char sequences like `{}` or `{:?}` - it will make things even less obvious.

Summing up: I think that white list is the best option. The only way to make it more flexible, as I see it, is to make white list configurable.

File: src/expr.rs (hunk 1 above)
Comment (nrc) [inline src/expr.rs]: Yes, that seems correct to me. I.e., it effectively ignores the first argument, and lays out the rest as if it didn't exist, other than causing a newline. In particular, it should follow any user preference for how the args are laid out, other than as they apply to the first one.

**Comment by nrc** [comment]
hmm, good point on the whitelisting...

I think I like the configurable whitelist option, where it should be included as one of the config options. By default, it can be the std ones. This means we don't need an option for this specifically. However, it will be a bit more work, sorry
